Also heres the interview question i had

IvarForkbeard opened the pull request:
Okay, let me know if I did this wrong.
Also let me know if I did this correctly.

Can't thank you guys enough for putting up with me!

IvarForkbeard committed:
Beginning work on the next challenge - making a branch to invite comment

IvarForkbeard committed:
My first official submission and maybe "pull request"?

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,31 @@
+let numberOfBuildings = streetArray.length;
+let tallestBuildingSoFar = 0;
+let bestArea = tallestBuilding * numberOfBuildings;
+let bannerOne = [tallestBuilding, numberOfBuildings];
+let bannerTwo = [tallestBuilding, numberOfBuildings];
+let bestBannerOne = [tallestBuilding, numberOfBuildings];
+let bestBannerTwo = [tallestBuilding, numberOfBuildings];
```
I would rather these broken out into their own individual variables

IvarForkbeard replied:
Can you elucidate?  You mean like, bannerOneTallestBuilding, bannerOneNumberOfBuildings, etc., so eight variables instead of four arrays?

danharasymiw replied:
`bannerOneHeight`/`width` would be appropriate I think

IvarForkbeard replied:
Oh, okay.  I can do that.  You're really gonna be unhappy that I added a new element to the array for area.  I'll variableifiy all those then.

eric-decompiled replied:
usually I use objects to keep related data together, like

```js
const banner1 = {tallest: tallestBuilding, numberOfBuildings: numberOfBuildings}
const banner2 = {tallest: tallestBuilding, numberOfBuildings: numberOfBuildings}
```

IvarForkbeard replied:
Nuts.  Who should I try to please?  Sensei Dan, or Sensei Eric?

danharasymiw replied:
If you want to use an object for it I'd say thats fine, I just wouldn't use an array to hold multiple values that represent different things altogether

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,31 @@
+        bestBannerTwo = [bannerTwo[0], bannerTwo[1]];
+    }
+}
+answer = "The best area we can cover is " + bestArea + " and that is from a banner of " + bestBannerOne[0] + "x" + bestBannerOne[1];
```
It's cool that you figured out how to log out the dimensions of the banner(s), but definitely not necessary. 
We just need to return the minimum sum of the banners that we were able to find. Ideally this means it's also in a function.

IvarForkbeard replied:
Oh, I suppose I agree, but I did want to "show my work" and I also thought it would be useful in debugging to know what banners my code was thinking.  I will deprecate those arrays.  Let me reflect on how I can meaningfully functionise something here.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,31 @@
+
+//    console.log("At address " + i + " the two banners required would be " + bannerOne + " and " + bannerTwo);
+
+    if (((bannerOne[0] * bannerOne[1]) + (bannerTwo[0] * bannerTwo[1])) < bestArea) {
```
nit: the extra parens actually aren't needed here, but if you find it more readable 🤷

IvarForkbeard replied:
Will I look old or feeble by having the extra brackets?  I do find it more readable and easier to follow the math blocks that way, but I could go through and deprecate them from code in future.  I actually even looked up if I could drop brackets from simple if statements altogether.  Spoiler: not in JS apparently, but it's okay in Python I think Themp said.

danharasymiw replied:
It's personal preference

eric-decompiled replied:
If you install prettier and have it run on save you it will take out extra brackets that aren't needed

https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

It stops most style related comments by automating it

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,31 @@
+const fs = require("fs");
+const streetArray = fs.readFileSync("data.txt", "utf-8").split(",");
+let answer = "";
+let tallestBuilding = Math.max(...streetArray);
+let numberOfBuildings = streetArray.length;
+let tallestBuildingSoFar = 0;
+let bestArea = tallestBuilding * numberOfBuildings;
+let bannerOne = [tallestBuilding, numberOfBuildings];
+let bannerTwo = [tallestBuilding, numberOfBuildings];
+let bestBannerOne = [tallestBuilding, numberOfBuildings];
+let bestBannerTwo = [tallestBuilding, numberOfBuildings];
```
nit: I'd personally prefer if there were some newlines in here to organize these a bit. It's personal preference though.

IvarForkbeard replied:
Hmmm, this does feel like will be very subjective, but l will try to read the tea leaves of what those who come after might prefer.

IvarForkbeard replied:
Marking this resolved as there wasn't a fresh complaint.  Prepared for that to change, as I might adjust these inits as per lower comments.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,31 @@
+        tallestBuildingSoFar = streetArray[i];
+    }
+    bannerOne = [tallestBuildingSoFar, i + 1];
+    bannerTwo = [Math.max(...streetArray.slice(i + 1, streetArray.length)), numberOfBuildings - i - 1];
```
This definitely works, but every time we shift the banner over, we have to recalculate the max height of the second banner. Is there a way to avoid this?

IvarForkbeard replied:
It feels like this isn't a rhetorical question.

danharasymiw replied:
;)

IvarForkbeard replied:
Have I addressed everything except this issue?

IvarForkbeard committed:
Made many Dan changes.

IvarForkbeard committed:
Broke arrays back into individual variables for readability.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,42 @@
+    
+    updateBanners(i);
+    
+    if (bannerOneArea+ bannerTwoArea < bestArea) {
```
nit: `bannerOneArea + bannerTwoArea`

IvarForkbeard replied:
After lots of finagling, I installed prettier as per @eric-decompiled - it fixed this for us.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,42 @@
+
+let bestArea = bannerOneArea;
+
+function updateBanners(address){
```
Address is fine I suppose, I personally prefer index if we're talking about positions in an array

IvarForkbeard replied:
This is something I'll have to reflect upon.  In this context, I thought address was pretty acceptable as it's analogous to us  walking down a street in front of the buildings.  Indeed, that's literally how the problem is described?

danharasymiw replied:
Is their position in the array their address? 🤷

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,42 @@
+let bannerOneArea = tallestBuilding * numberOfBuildings;
+let bannerTwoHeight = tallestBuilding;
+let bannerTwoWidth = numberOfBuildings;
+let bannerTwoArea = tallestBuilding * numberOfBuildings;
```
Should we start off with two banners initialized to cover the max area? I would've expected only one

IvarForkbeard replied:
I wanted to initialise these values to the absolute worst case scenario to ensure that they would get updated right out of the gate and not potentially poison any final results.  Let me reflect on if they need to be initialised with any value really.

danharasymiw replied:
You know, I guess it really doesn't matter.
Worst possible case though is still one banner covering everything 🤷

danharasymiw reviewed (commented):
I'm glad we have a function now, although it's not quite what I would've expected.

I think you should try and have a function like:
```js
func findMinArea(buildings)
```
And then you'd set your `answer` string like so:
```
answer = "The best area we can cover is  " + findMinArea(buildings);
```

IvarForkbeard committed:
Minor cleanup.  Still puzzling out not recalculating every shift.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,43 @@
+    return a - b;
+});
+
+let numberOfBuildings = streetArray.length;
```
Since these variables are only referenced inside of our `findBestArea` function, maybe we should move them there?

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,43 @@
+let bannerTwoWidth = 0;
+let bannerTwoArea = 0;
+
+function findBestArea() {
```
I'd expect this function to accept the array of building heights, that way we can reuse this function

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,43 @@
+let bannerTwoArea = 0;
+
+function findBestArea() {
+    let bestAreaInFunction = bannerOneArea;
```
maybe just `bestArea`? or `minArea` (since thats what we're aiming for)

IvarForkbeard committed:
Hopefully getting closer, still a concept to grasp.

IvarForkbeard committed:
Slightly optimised.  Won't check max remaining EVERY time.

IvarForkbeard commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        if ((streetArray[i] = tallestBuildingRemaining)) {
+            tallestBuildingRemaining = Math.max(...streetArray.slice(i + 1, streetArray.length));
```
Okay, so I added this to appease the lesser gods.  Now it only bothers checking what's the tallest building remaining on the street if the current building is part of the group known as the tallest building remaining.  That way if it IS the only member of tallest building remaining, we'll snag whatever building is tallest from the smaller group.  But from then on, it's not checking them, it just remembers what the tallest building is until it actually encounters one of that height.

eric-decompiled reviewed (approved):
lgtm

IvarForkbeard committed:
Fulfilled requirements as set out.

IvarForkbeard committed:
Removed Eric approved version for clarity.

IvarForkbeard committed:
Renaming this to my solution, waiting for feedback, will push to main

Then I'll maybe try to code up a solution similar to what was intended.  But I sure didn't want to lose this one as I poured my heart and soul into it.

IvarForkbeard committed:
How I would have done it last night.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+
+function bestArea(arrStreet) {
+    let numberOfBuildings = arrStreet.length;
+    let staHeights = new Array(numberOfBuildings);
```
I feel like I only know `sta` is for a stack given our convo thus far.
I'm amused we tried to save 2 characters from this name and have other really long variable names lol
I'd probably just name this `heightStack`

danharasymiw replied:
Stacks should also start off empty....
but this works perfectly well for a list of `maxHeightAtIndex`. Totally acceptable as well, I'd just probably update the name to match, or keep the same name and actually use it like a stack

IvarForkbeard replied:
Why is arr okay, but sta invites scrutiny?  I tried to google if there was a convention, but there didn't seem to be much of one...
I was finding in JS, that if I compared the current value of the stack with a number, and the element of the stack didn't actually exist, it said that NaN was greater than the actual number.  so I decided to prepopulate the stack with naughts to make sure that didn't happen.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+    let bannerRightWidth = 0;
+    let bannerRightHeight = 0;
+    let bannerRightArea = 0;
+    let bestAreaInFunction = 0;
```
I'd rename this to just `bestArea`

IvarForkbeard replied:
the function itself is called bestArea - wouldn't that invite confusion at the very least?

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+        bannerLeftArea = bannerLeftWidth * bannerLeftHeight;
+        bannerRightArea = bannerRightWidth * bannerRightHeight;
+
+        if (bestAreaInFunction === 0 || bannerLeftArea + bannerRightArea < bestAreaInFunction) {
```
We shouldn't need this `=== 0` check, we have all the information we need to calculate a default value for the `bestArea` before this for loop :)

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+
+    for (let i = numberOfBuildings - 1; i > 0; i--) {
+        bannerRightHeight = Math.max(bannerRightHeight, arrStreet[i]);
+        bannerRightWidth = Math.abs(i - numberOfBuildings);
```
This works... but can't we just do `numberOfBuildings - i` and just avoid the negative numbers altogether?

IvarForkbeard commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+    let bestAreaInFunction = 0;
+
+    for (let i = 0; i < numberOfBuildings; i++) {
+        tallestBuildingSoFar = Math.max(tallestBuildingSoFar, arrStreet[i]);
```
While Math.max was originally prohibited by Dan, he used max in his Python example and I feel that my solution here is using the same spirit of the comparison as he used in his own example, rather than me using this to analyse the contents of the entire array.  This same technique is employed below.  If it is an issue, I am confident I can achieve the same outcome with an if/then statement.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHadMySolution.js:
```diff
@@ -0,0 +1,96 @@
+    return bestAreaInFunction;
+}
+
+bestArea(buildings);
```
don't need this

IvarForkbeard replied:
You already approved of this code?  It's my original solution, and I thought it wasn't even up for further discussion until Eric had picked at it?

danharasymiw replied:
I'm not perfect :( but I am working on it

IvarForkbeard committed:
Addressing all of Dan's nits.

danharasymiw commented on alsoHeresTheInterviewQuestionIHad/alsoHeresTheInterviewQuestionIHad.js:
```diff
@@ -0,0 +1,89 @@
+    let bannerRightWidth = 0;
+    let bannerRightHeight = 0;
+    let bannerRightArea = 0;
+    let bestArea = 0;
```
nit: Remove this and then down below you could do
```js
let bestArea = numberOfBuildings * bannerLeftHeight
```
instead of setting it's value twice

IvarForkbeard committed:
Minor fixes, added a check to the output.

IvarForkbeard committed:
Updated My Version to also have a desired output.

danharasymiw reviewed (approved):
I think you could just make it so that all the test data lives in the same file, each on it's own line but, you've solved the problem twice now 🤷‍♂️ 

+1
